Introduce query-by-tuple syntax

paarthmadan opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This PR is related to the ongoing effort of introducing composite primary keys.

A very common Rails abstraction pattern is performing a search given a primary key. Across Rails we see numerous instances of code like this:

- [Example 1](https://github.com/rails/rails/blob/3dd5d4900d01c0b25f44c2fbf6ce261d367a490a/activerecord/lib/active_record/destroy_association_async_job.rb#L26)
- [Example 2](https://github.com/rails/rails/blob/fdceee8d78711dff3af24394dda978efe279c084/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L1368)
- [Example 3](https://github.com/rails/rails/blob/d53dc868869f973c7ee0bbcdaacfc0239f261773/activerecord/lib/active_record/relation/finder_methods.rb#L404)

Now that a model can have a primary key that's composed of more than 1 column, a common querying pattern is to look for a set of records given their composite keys.

This PR brings support for querying using tuple-syntax (happy to hear suggestions on other names):

```ruby
book_one = Cpk::Book.create!(author_id: 1, number: 2)
book_two = Cpk::Book.create!(author_id: 3, number: 4)

# Provide a mapping between a list of columns and an array of corresponding tuples
Cpk::Book.where([:author_id, :number] => [[1, 2]]) # Relation(book_one)
Cpk::Book.where(Cpk::Book.primary_key => [[1, 2], [3, 4]]) # Relation(book_one, book_two)
```

### Detail

This PR changes the predicate builder to consider the case when a key in the `Hash` supplied to `#where` is an array. In that case, it iterates across the set of tuples and computes what the query would look like individually. Previously, it was assumed keys and values were singular. Now, the method accepts an array. When encountering one, we iterate through each tuple, and compute a query by deferring each sub-task to the original implementation (i.e recursively), and then finally regrouping (collecting) the queries.

The implementation of `grouping_queries` injects an `OR` in between sub-queries.

Note: This PR doesn't concern itself too heavily with the performance or format of the generated SQL. We've started to consider future iterations where a user-defined config (think: `active_record.config.bulk_query_strategy`) can be used to configure how and what SQL Active Record generates to perform this query. This would allow us to consider building row constructor SQL, appending hints in sharded contexts, or simply falling back to the current "OR-separated" clauses.

In any case, this PR only concerns itself with designing the API for querying by tuple and is less focused on how the query executes.

This PR will help avoid branching and commits like 26994cb06e6214bf387cc34adbb7fb29834b715e will soon be common as we replace many instances of branching with this new syntax.

This new syntax also helps de-risk the composite primary key rollout because many callsites will "automatically" work now that the abstraction works in both the singular and composite case.

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

paarthmadan committed:
Specify where clauses by mapping columns to tuples

With the introduction of composite primary keys, a common usecase is querying for records with tuples representing the composite key. This change introduces new syntax to the where clause that allows specifying an array of columns mapped to a list of corresponding tuples. It converts this to an OR-joined list of separate queries, similar to previous implementations that rely on grouping queries.

paarthmadan committed:
Use tuple query syntax to avoid branching

paarthmadan commented:
Thoughts on this change requiring a CHANGELOG entry? I'd assume it warrants one because it introduces a pretty significant syntax change that's not directly tied to CPK. Thoughts: cc @eileencodes @nvasilevski

eileencodes commented:
And yes please add a changelog entry 👍🏼

eileencodes commented:
@paarthmadan - intersting, the build is failing but not consistently. Looks like one of the new tuple tests is flaky. Can you take a look? https://buildkite.com/rails/rails/builds/94980#01870a78-1d3b-4bb6-ad35-e859d3597858/1036-1046

paarthmadan committed:
Delete Cpk::Book in non-transactional test

Having this record persisted across test runs was leading to flakey behaviour in tests that were creating Cpk::Book with the same primary key.
